backend/src/LibExecution/ProgramTypes.fs

```diff
@@ -152,21 +165,68 @@ type Expr =
   | ETuple of id * Expr * Expr * List<Expr>
   | ERecord of id * List<string * Expr>
   | EPipe of id * Expr * Expr * List<Expr>
+
   // Constructors include `Just`, `Nothing`, `Error`, `Ok`.  In practice the
   // expr list is currently always length 1 (for `Just`, `Error`, and `Ok`)
   // or length 0 (for `Nothing`).
+  // TODO: migrate usages of this to usages of EDefinedEnum(FQTypeName.T, ...
   | EConstructor of id * string * List<Expr>
-  | EMatch of id * Expr * List<MatchPattern * Expr>
+
+  /// Supports `match` expressions
+  /// ```fsharp
+  /// match x + 2 with // arg
+  /// // cases
+  /// | pattern -> expr
+  /// | pattern -> expr
+  /// | ...
+  /// ```
+  | EMatch of id * arg : Expr * cases : List<MatchPattern * Expr>
+
   // Placeholder that indicates the target of the Thread. May be movable at
   // some point
   | EPipeTarget of id
-  // EFeatureFlag: id, flagName, condExpr, caseAExpr, caseBExpr
-  | EFeatureFlag of id * string * Expr * Expr * Expr
+
+  /// Like an if statement, but with a label
+  /// TODO: continue describing
+  | EFeatureFlag of
+    id *
+    flagName : string *
+    cond : Expr *
+    caseA : Expr *
+    caseB : Expr
+
+
+  // TODO:
+  // - define EUser
+  // migrate existing ERecords to EUserRecords and
+  // /// Given a User type of:
+  // ///   `type MyRecord = { A: int;  B: int * MyRecord }`
+  // /// , this is the expression
+  // ///   `EUserRecord(UserType.MyRecord, [EInteger(1), EString("title")]`
+  // | EUserRecord of id * UserTypeName * fields: List<string * Expr>
+
+
+  // TODO one of these:
+  // - implement EStdlibEnum and EStdlibRecord, then EPackageEnum and EPackageRecord
+  // - implement a more generic EDefinedEnum and EDefinedRecord
+  //   that reference awith `User`, `Stdlib`, and `Package` cases
+
+  /// Given a User type of:
+  ///   `type MyEnum = A | B of int | C of int * (label: string) | D of MyEnum`
+  /// , this is the expression
+  ///   `EUserEnum(UserType.MyEnum, "C", [EInteger(1), EString("title")]`
+  | EUserEnum of id * UserTypeName * caseName : string * fields : List<Expr>
```

Comment:
This should be integrated with EConstructor, right?

This isn't an enum, afaik. This is a constructor (the enum is the type definition, right?)

Reply:
You're right, it is an EConstructor. I can get those merged together.

I've had a really hard time with the name "constructor" not referring to something in OOP space - would you be open to EEnumCtor or something else, that's not plainly "EConstructor"?

Reply:
some other ideas, some provided by chatgpt: EVariant, EUnionCase, ECase. If we can't find something that works, I'd be fine sticking with EConstructor. It's just been bugging me for a while for some reason. F#'s pattern to extract the values out of enums in `match` patterns is called the "identity" pattern if that's worth anything